src/tools/base64.cpp

```diff
@@ -1,172 +1,117 @@
+{
+    if (inputString == NULL || len == 0 ) {
+        spdlog::error("Invalid data");
+        return std::string();
+    }
 
-#include "../tools/base64.h"
-
-#include <cstring>
-
-/* aaaack but it's fast and const should make it shared text page. */
-static const unsigned char pr2six[256] = {
-    /* ASCII table */
-    64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
-    64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
-    64, 64, 64, 64, 64, 62, 64, 64, 64, 63, 52, 53, 54, 55, 56, 57, 58, 59, 60,
-    61, 64, 64, 64, 64, 64, 64, 64, 0,  1,  2,  3,  4,  5,  6,  7,  8,  9,  10,
-    11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 64, 64, 64, 64,
-    64, 64, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42,
-    43, 44, 45, 46, 47, 48, 49, 50, 51, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
-    64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
-    64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
-    64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
-    64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
-    64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
-    64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
-    64, 64, 64, 64, 64, 64, 64, 64, 64};
-
-size_t Base64decode_len(const char* bufcoded) {
-  size_t nbytesdecoded;
-  const unsigned char* bufin;
-  int nprbytes;
-
-  bufin = reinterpret_cast<const unsigned char*>(bufcoded);
-  while (pr2six[*(bufin++)] <= 63)
-    ;
-
-  nprbytes = (bufin - (const unsigned char*)bufcoded) - 1;
-  nbytesdecoded = ((nprbytes + 3) / 4) * 3;
-
-  return nbytesdecoded + 1;
-}
+    std::string opString;
+    opString.reserve((len + 2) / 3 * 4);
+    const uint8_t* base64chars = B64chars;
+    unsigned int pos = 0;
+    constexpr int encodeWordLen = 3;
+
+    while (pos < len) {
+        opString.push_back(base64chars[(inputString[pos] & 0xfc) >> 2]);
+
+        if (pos+1 < len) {
+            opString.push_back(base64chars[((inputString[pos] & 0x03) << 4) + ((inputString[pos + 1] & 0xf0) >> 4)]);
+
+            if (pos+2 < len) {
+                opString.push_back(base64chars[((inputString[pos + 1] & 0x0f) << 2) + ((inputString[pos + 2] & 0xc0) >> 6)]);
+                opString.push_back(base64chars[inputString[pos + 2] & 0x3f]);
+            }
+            else {
+                opString.push_back(base64chars[(inputString[pos + 1] & 0x0f) << 2]);
+                opString.push_back('=');
+            }
+        }
+        else {
+            opString.push_back(base64chars[(inputString[pos] & 0x03) << 4]);
+            opString.push_back('=');
+            opString.push_back('=');
+        }
+
+        pos += encodeWordLen;
+    }
+
+    return opString;
+};
+
+std::string base64::base64decode(std::string_view encodedString, const size_t &len)
```

Comment:
why do you pass the length as a parameter ? you can check the length by  encodedString.Length()

Reply:
![image](https://github.com/eclipse-uprotocol/uprotocol-sdk-cpp/assets/145686444/0f78684a-4a18-4928-b2bd-bfa0a8e5f837)

In this case, we don't need to overload the "base64decode(std::string_view encodedString, const size_t &len)" and pass the parameter length, below two functions are enough to fulfill requirements.

static std::string base64encode(std::string  t_str);
static std::string base64decode(std::string  t_str);

Reply:
static std::string base64encode(uint8_t const* data, const size_t &len);
static std::string base64decode(uint8_t const* data, const size_t &len);

static std::string base64encode(std::string str);
static std::string base64decode(std::string str)

These are the API`s that you should have

Reply:
static std::string base64decode(**uint8_t const* data**, const size_t &len);
Can you please confirm which data type we have to use "string" or "uint8_t" for the data argument?